src/groups/general_unitary_groups.jl

```diff
@@ -0,0 +1,262 @@
+        end
+        a₀ = (α * sinα + (1 + r - d) * cosα) * c
+        a₁ = ((3 - d) * sincα - (2 - r) * cosα) * c
+        a₂ = (sincα - (1 - r) / 2 * cosα) * c
+        a₃ = (e + (1 - r) * (e - sincα / 2)) * c
+    end
+
+    X² = X * X
+    X³ = X² * X
+    q = a₀ * I + a₁ .* X .+ a₂ .* X² .+ a₃ .* X³
+    return q
+end
+
+inverse_translate(G::GeneralUnitaryMultiplicationGroup, p, q, ::LeftAction) = inv(G, p) * q
+inverse_translate(G::GeneralUnitaryMultiplicationGroup, p, q, ::RightAction) = q * inv(G, p)
+
+function inverse_translate!(G::GeneralUnitaryMultiplicationGroup, x, p, q, ::LeftAction)
+    return mul!(x, inv(G, p), q)
+end
+function inverse_translate!(G::GeneralUnitaryMultiplicationGroup, x, p, q, ::RightAction)
+    return mul!(x, q, inv(G, p))
+end
+
+function inverse_translate_diff(
+    G::GeneralUnitaryMultiplicationGroup,
+    p,
+    q,
+    X,
+    conv::ActionDirection,
+)
+    return translate_diff(G, inv(G, p), q, X, conv)
+end
+function inverse_translate_diff!(
+    G::GeneralUnitaryMultiplicationGroup,
+    Y,
+    p,
+    q,
+    X,
+    conv::ActionDirection,
+)
+    return copyto!(Y, inverse_translate_diff(G, p, q, X, conv))
+end
+
+function log_lie!(
+    G::GeneralUnitaryMultiplicationGroup{n,ℝ},
+    X::AbstractMatrix,
+    q::AbstractMatrix,
+) where {n}
+    log_safe!(X, q)
+    return project!(G, X, Identity(G), X)
+end
+function log_lie!(
+    ::GeneralUnitaryMultiplicationGroup{n,ℝ},
+    X,
+    ::Identity{MultiplicationOperation},
+) where {n}
+    fill!(X, 0)
+    return X
+end
+function log_lie!(
+    G::GeneralUnitaryMultiplicationGroup{2,ℝ},
+    X::AbstractMatrix,
+    q::AbstractMatrix,
+)
+    @assert size(q) == (2, 2)
+    @inbounds θ = atan(q[2, 1], q[1, 1])
+    return get_vector!(G, X, Identity(G), θ, DefaultOrthogonalBasis())
+end
+function log_lie!(
+    G::GeneralUnitaryMultiplicationGroup{3,ℝ},
+    X::AbstractMatrix,
+    q::AbstractMatrix,
+)
+    e = Identity(G)
+    cosθ = (tr(q) - 1) / 2
+    if cosθ ≈ -1
+        eig = eigen_safe(q)
+        ival = findfirst(λ -> isapprox(λ, 1), eig.values)
+        inds = SVector{3}(1:3)
+        ax = eig.vectors[inds, ival]
+        return get_vector!(G, X, e, π * ax, DefaultOrthogonalBasis())
+    end
+    X .= q ./ usinc_from_cos(cosθ)
+    return project!(G, X, e, X)
+end
+function log_lie!(
+    G::GeneralUnitaryMultiplicationGroup{4,ℝ},
+    X::AbstractMatrix,
+    q::AbstractMatrix,
+)
+    cosα, cosβ = cos_angles_4d_rotation_matrix(q)
+    α = acos(clamp(cosα, -1, 1))
+    β = acos(clamp(cosβ, -1, 1))
+    if α ≈ 0 && β ≈ π
+        A² = Symmetric((q - I) ./ 2)
+        P = eigvecs(A²)
+        E = similar(q)
+        fill!(E, 0)
+        @inbounds begin
+            E[2, 1] = -β
+            E[1, 2] = β
+        end
+        copyto!(X, P * E * transpose(P))
+    else
+        det(q) < 0 && throw(
+            DomainError(
+                "The LIe group logarithm is not defined for $q with a negative determinant ($(det(q)) < 0).",
```

Comment:
```suggestion
                "The Lie group logarithm is not defined for $q with a negative determinant ($(det(q)) < 0). Point `q` is in a different connected component of the manifold $G",
```
A bit more detailed explanation.